src/portable/st/synopsys/dcd_synopsys.c

```diff
@@ -1040,6 +1114,210 @@ void dcd_int_handler(uint8_t rhport)
     // IEPINT bit read-only
     handle_epin_ints(rhport, dev, in_ep);
   }
+
+  // Check for Incomplete isochronous IN transfer
+  if(int_status & USB_OTG_GINTSTS_IISOIXFR) {
+    TU_LOG2("      IISOIXFR!\r\n");
+  }
+}
+
+// Helper function which parses through the current configuration descriptors to find the biggest EPs in size.
+static bool get_ep_size_report(uint8_t rhport, tusb_desc_configuration_t const * desc_cfg, ep_sz_tt_report_t * p_report)
+{
+  (void) rhport;
+
+  //  tu_memclr(p_report, sizeof(ep_sz_tt_report_t));   // This does not initialize the first two entries ... i do not know why!
+
+  // EP0 sizes and usages are fixed
+  p_report->ep_size[0][TUSB_DIR_OUT] = p_report->ep_size[0][TUSB_DIR_IN] = CFG_TUD_ENDPOINT0_SIZE;
+  p_report->ep_transfer_type[0][TUSB_DIR_OUT][TUSB_XFER_CONTROL] = p_report->ep_transfer_type[0][TUSB_DIR_IN][TUSB_XFER_CONTROL] = true;
+
+  // Parse interface descriptor
+  uint8_t const * p_desc   = ((uint8_t const*) desc_cfg) + sizeof(tusb_desc_configuration_t);
+  uint8_t const * desc_end = ((uint8_t const*) desc_cfg) + desc_cfg->wTotalLength;
+
+  uint8_t addr;
+
+  while( p_desc < desc_end )
+  {
+    if (TUSB_DESC_ENDPOINT == tu_desc_type(p_desc))
+    {
+
+      addr = ((tusb_desc_endpoint_t const*) p_desc)->bEndpointAddress;
+
+      // Verify values - this checks may be omitted in case we trust the descriptors to be okay
+      TU_VERIFY(tu_edpt_number(addr) < EP_MAX);
+      TU_VERIFY(tu_edpt_dir(addr) <= TUSB_DIR_IN);
+      TU_VERIFY(((tusb_desc_endpoint_t const*) p_desc)->bmAttributes.xfer <= TUSB_XFER_INTERRUPT);
+
+      p_report->ep_size[tu_edpt_number(addr)][tu_edpt_dir(addr)] = tu_max16(p_report->ep_size[tu_edpt_number(addr)][tu_edpt_dir(addr)], ((tusb_desc_endpoint_t const*) p_desc)->wMaxPacketSize.size);
+      p_report->ep_transfer_type[tu_edpt_number(addr)][tu_edpt_dir(addr)][((tusb_desc_endpoint_t const*) p_desc)->bmAttributes.xfer] = true;
+    }
+    p_desc = tu_desc_next(p_desc); // Proceed
+  }
+  return true;
+}
+
+// Setup FIFO buffers at configuration time.
+// The idea is to use this information such that the FIFOs need to be configured only once
+// at configuration time and no more later. This makes it easy in case you want to close and open EPs for different
+// purposes at any time (without taking care of other active EPs). You also avoid the nasty need of defragmenting
+// the TX buffers, which is likely to happen.
+// Certainly, this function does not allow for the highest possible flexibility as it only works in the worst case
+// (all biggest EPs can be active at the same time). However, this should not be a problem for almost all applications.
+
+// Spare space is assigned equally divided to bulk and interrupt EPs.
+// Pure isochronous EPs do not get any spare space as it does not make any sense.
+TU_ATTR_WEAK bool dcd_alloc_mem_for_conf(uint8_t rhport, tusb_desc_configuration_t const * desc_cfg)
```

Comment:
To be honest, I don't like the introduction of this function. defragmenting is not a real issue, it is mostly switching between open/close ISO endpoint, which could be addressed in other way like introducing `dcd_edpt_iso_open` with additional max packet size of all alternate Itf. If you could revert this change and still having the audio_test example running. I will update the ISO and dcd to better support open/close endpoint.

Reply:
I see your reason why it would be better to have such a configuration parsing within USBD and not in DCD. This totally makes sense since other MCUs may need the same. However, i am not sure if handling ISO EPs without taking care of the remaining EPs is sufficient! Your idea of reserving space only for the ISO EPs is similar to my approach but i also take care that there is space for all used remaining EPs. So an overflow can be avoided and detected at compilation time. Every EP has its dedicated space available so you also do not need any memory management.
The very best thing would be to be able to close EPs and defragment the FIFO right away. In this case you would not need any pre-FIFO-allocation scan at all. But i am not sure if this is easily done on the fly.

If you insist i certainly revert the allocation scheme but i think it would be worth considering the alternatives.

Reply:
>  Your idea of reserving space only for the ISO EPs is similar to my approach but i also take care that there is space for all used remaining EPs.

No I didn't say we will reserve space for ISO only. First of all the only endpoint that get closed and reopened is only ISO. It is unspoken truth that for all the ISO endpoints, its default interface (alternate = 0) must has no endpoint (closed ISO). Therefore ISO will likely be the last endpoints got opened. All other endpoints are opened when device receives **SET_CONFIGURED** by their driver. The ISO will need an additional **SET_INTERFACE** to actually open the atl one with corresponding packet size.

What I am trying to say is we don't have to deal with it like malloc() when open()/close() randomly, only few last endpoints are closed and re-open. When the usbd/class driver go through all the Alt Interface (e.g Audio streaming), It will know the max size of all the ATL e.g Atl 1 has size = 100, atl 2 size = 125 . The first time it open the endpoint even though with size = 100, it will also pass the max size of 125 so that dcd can reserve that space for easy close/reopen. If the FIFO size does not afford the max size of 125 for ISO, the description shouldn't include that Alt with 125 (should only support 100 e.g).

Yeah, sure it needs a bit more work now which I will be happy to do later on, however I think any higher level than endpoint should be decided by USBD/class driver.

Reply:
Ah now the picture gets clearer. Actually a smart thought! Your are right, the interfaces which need to be activated by **SET_INTERFACE** will use what is left from the other EPs. I always have this highly complicated picture in mind, where the end user may has multiple multi-interface USB applications running (similar to audio). I guess this comes from reading too many USB specifications ^^. They always suggest you may close interfaces currently not needed in order to allow a bigger interface to be opened and so on.

I can see that your approach works definitely if you have only one audio interface running. In case you have two such interfaces you need additional logic to recognize if a FIFO was already setup with a bigger space (e.g. 125). This is the memory management i meant before.

To summarize you and i try to solve the very same problem in a very similar manner. Actually you definitely too reserve space but only for ISO EPs.

> The first time it open the endpoint even though with size = 100, it will also pass the max size of 125 so that dcd can reserve that space for easy close/reopen.

This brings the need for memory management and additional APIs for ISO EPs. My approach parses everything at the beginning, sets the FIFO once and does not change anything later on. So there is no memory management needed but it needs to be done once a new configuration gets set and can not be done by a call to ep_open().

I suggest i revert my approach for now. If desired, after all it is your choice at the end, we could introduce this in a separate PR.

Reply:
> Ah now the picture gets clearer. Actually a smart thought! Your are right, the interfaces which need to be activated by **SET_INTERFACE** will use what is left from the other EPs. I always have this highly complicated picture in mind, where the end user may has multiple multi-interface USB applications running (similar to audio). I guess this comes from reading too many USB specifications ^^. They always suggest you may close interfaces currently not needed in order to allow a bigger interface to be opened and so on.

Yeah, USB docs is over-engineer for sure, but they do a really good job with 2.0. The term after 3.0/3.1 gen1 gen2 are too messy :laughing: 

> This brings the need for memory management and additional APIs for ISO EPs. My approach parses everything at the beginning, sets the FIFO once and does not change anything later on. So there is no memory management needed but it needs to be done once a new configuration gets set and can not be done by a call to ep_open().  
 
We may need one but it is not really complicated. I think for stm32, we only need to update the fifo index with a smaller size, there is probably no need to shrink + combine fifo together. Actually, your implementation still need to handle this case of  SET_INTERFACE request for changing the endpoint size as well e.g in video class the ISO endpoint size is determined by resolution in Video Streaming Data
- ALT 0 : no endpoint
- ATL 1: 800 x 600 : size = 200
- ATL 2: 1024 x 768 : size = 250
- ATL 3: 1280 x 720 : size = 300

And this switch dynamically within an same configuration descriptor depending on the user selection of the resolution. What I propose is the class driver take note of the max possible ATL = 300, then pass/set it to reserve the space, from there it could switch back and forth with ease (only change the fifo size) and it shouldn't affect the index and space of the endpoint after it (multiple audio or video + audio).

> I suggest i revert my approach for now. If desired, after all it is your choice at the end, we could introduce this in a separate PR.

I am glad to hear you think my approach make sense. Yeah, I think if needed, we could have this a its own PR. Once we have everything working (now It should since the ISO is the last), and we can work from there for optimization. 

PS: The next PR I will love to work on is getting the `audio_speaker` or `audio_microphone` so that we could visually see or rather hear the actual result.

Reply:
> Actually, your implementation still need to handle this case of SET_INTERFACE request for changing the endpoint size as well...

Thats acutally not correct, my implementation would recognize ATL 3 and reserve the required 300 in the first place. It determines the worst case i.e. the biggest required FIFO size for every EP ever needed and sets up the FIFO this way. Thus the FIFO never needs to be reconfigured even in case of open/close EPs and it does not need any management.

Regarding the examples, I can update the microphone example but a speaker example would best be given by @kasjer since he already tested the speaker case.

Reply:
Both microphone and speaker in real life require code that is capable of handling sound.
This is very hardware dependent and I don't think we can have out of the box code that will work on more then one platform.
For microphone we can have some sort of noise generator that will produce stream that host can see.
For speaker it is a little more difficult, audio data has to go somewhere.

For my part I'm using I2S microphone and I2S speaker that works with STM32F4 MCUs.
Speaker also works with NRF52 (microphone is not working due to NRF52 I2S 24 bits limitation, microphone accepts only LRCL=BCLK/64).
Speaker also works with DA1469x (16 bits stereo) (microphone not done yet due to I2S not being ready in mynewt).

I though that I will do headset instead that will have both microphone and speaker at the same time.
That way we can have an example that does not need any specific hardware and will simply route speaker 
to microphone (stereo to mono).

This partially works on Linux, Windows rejects my headset device, Mac I did not checked yet.

Reply:
> Thats acutally not correct, my implementation would recognize ATL 3 and reserve the required 300 in the first place. It determines the worst case i.e. the biggest required FIFO size for every EP ever needed and sets up the FIFO this way. Thus the FIFO never needs to be reconfigured even in case of open/close EPs and it does not need any management.

Yeah, I mean I know you reserve the largest space for ISO. But we still need to resize the ep when SET_INTERFACE with different ep size is received. For stm32, maybe we can leave it with max fifo size and should be no issues, other mcus may need different approach. I kind of mixed-up some thought along the way across MCUs approach.
 
> Regarding the examples, I can update the microphone example but a speaker example would best be given by @kasjer since he already tested the speaker case.

> Both microphone and speaker in real life require code that is capable of handling sound.
> This is very hardware dependent and I don't think we can have out of the box code that will work on more then one platform.
> For microphone we can have some sort of noise generator that will produce stream that host can see.
> For speaker it is a little more difficult, audio data has to go somewhere.
> 
> For my part I'm using I2S microphone and I2S speaker that works with STM32F4 MCUs.
> Speaker also works with NRF52 (microphone is not working due to NRF52 I2S 24 bits limitation, microphone accepts only LRCL=BCLK/64).
> Speaker also works with DA1469x (16 bits stereo) (microphone not done yet due to I2S not being ready in mynewt).
> 
> I though that I will do headset instead that will have both microphone and speaker at the same time.
> That way we can have an example that does not need any specific hardware and will simply route speaker
> to microphone (stereo to mono).
> 
> This partially works on Linux, Windows rejects my headset device, Mac I did not checked yet.

Superb !!! Thank you very much, that would be great to have a visually working example makes it super easy when refactoring/enhancing let's say using DMA fifo/ EP allocation scheme :).  There is no rush though, and we don't need to have a complete portable example. Not all the boards has built-in mic/speaker anyway, and I am totally agreed with the noise/sound generator that would make the example runs on most of the boards. The speaker is optional, we can start with specific supported board, then move up later on. I could help with that :).

Reply:
> Yeah, I mean I know you reserve the largest space for ISO. But we still need to resize the ep when SET_INTERFACE with different ep size is received. For stm32, maybe we can leave it with max fifo size and should be no issues, other mcus may need different approach. I kind of mixed-up some thought along the way across MCUs approach.

That is correct, the EP needs to know its current size but the FIFO size is fixed to maximum required size so there will never occur any issues once everything was setup. I also don't know how other MCUs work here, i only use STMs. I am reverting the UAC2 driver and we handle this in a separate PR. Some additional time to allow creativity hammering out the best approach is never a bad idea! ;)

For testing purpose, currently i am testing my application by sending a saw-tooth signal (i just send increasing values until it overflows and so forth). This does not sound nice but one can check visually (e.g. via Python or other software capturing and plotting a sound snippet) if there are no missing values. You may do the same for speaker and check via a debugger the content of the receive FIFO. Those are manual testing procedures i know. The noise and especially the head-set example are a brilliant idea!